### ISSUE TITLE
Quickstart Tutorial Different Results

### ISSUE THREAD

**Issue by Kyroba** [loganbvh/py-tdgl#26]
First of all, what a wonderfully simple to use package. Thank you very much for designing this.

The issue I am experiencing is that the results are slightly different when running the exact same Quickstart notebook to what is shown in (https://py-tdgl.readthedocs.io/en/latest/notebooks/quickstart.html). I believe this might be a result of the mesh generation, was there a change in how it is done? I noticed that the tutorial uses tdgl version 0.4.0 whereas I am using the latest release 0.3.1.

What hardware was used to achieve those computation times in the quickstart tutorial?

I have attached my notebook results for more information if needed. 
PC specs: i9 9900k processor, 2080 ti gpu, 32 gb ram 3200Mhz, 980 PRO NVME SSD.

[JupyterLab_Quickstart.pdf](https://github.com/loganbvh/py-tdgl/files/12482089/JupyterLab_Quickstart.pdf)

**Comment by Kyroba** [comment]
4 Minutes after I posted that 0.4.0 released,

[JupyterLab_Quickstart0_4_0.pdf](https://github.com/loganbvh/py-tdgl/files/12482150/JupyterLab_Quickstart0_4_0.pdf)

However, no changes to my outputs. Not sure what it could possibly be.

**Comment by loganbvh** [comment]
Hi @Kyroba,

I just released v0.4.0 to pypi https://pypi.org/project/tdgl/0.4.0/. I meant to release it several weeks ago, but just forgot.

The results on https://py-tdgl.readthedocs.io/en/latest/ were obtained on my M1 MacBook Pro. There can be slight differences in meshing on different hardware and operating systems. For example, the quickstart notebook generates a mesh with 4633 sites on my MacBook Pro, whereas on [Google Colab](https://colab.research.google.com/github/loganbvh/py-tdgl/blob/main/docs/notebooks/quickstart.ipynb) (which is probably a Linux OS, not sure about the CPU), the same notebook generates a mesh with 4716 sites. On your Windows/Intel machine, the number of mesh sites is also slightly different (4756 sites). I am not sure why this happens, but it probably has to do with the [meshpy](https://documen.tician.de/meshpy/) library, which uses the [Triangle](http://www.cs.cmu.edu/~quake/triangle.html) C library. 

It's not surprising that differences in meshing like this can result in sub-percent changes in the simulation results, so I think there is no problem. Of course, it would be preferable if results were exactly the same on all platforms, but that's not always the case with numerical simulations.

Please let me know if you have any other issues or questions.

**Comment by Kyroba** [comment]
Thank you for the explanation! The only other issue I was having with the notebook was the animation calls. I realised I needed to install FFmpeg and include in path. The first animation with zero applied field took 6 minutes to make. Is this similar for you? 

I was also wondering why the computation time (thermalizing and simulating) on my end is almost double? Have you done much testing on windows systems? I have tried it on my other pc with an i9 12900k and the performance was somewhere between my i9 9900k and your macbook. I can try and test some other devices I have to compare.

Before I try to simulate larger and more complex systems I would first like to reduce any bottlenecks from my end. Any ideas?

**Comment by loganbvh** [comment]
Yes, generating animations in the Jupyter notebooks takes a long time - generally longer than the actual simulation takes. To quickly visualize results, you can use the `tdgl.visualize` command line tool: https://py-tdgl.readthedocs.io/en/latest/api/visualization.html#cli-tool-tdgl-visualize.

I have not done much testing on Windows. One thing that you could try is to uninstall scipy using pip (`pip uninstall scipy`) and then re-install scipy using the conda package manager (`conda install scipy`). You could also try the same for numpy. I think that on Windows the conda version of these packages may use more highly-optimized numerical libraries (e.g. Intel MKL https://en.wikipedia.org/wiki/Math_Kernel_Library).

**Comment by Kyroba** [comment]
Thank you! I am now using the conda versions of both, and the mesh generation has changed to 4627 sites, but the computation speed for all calculations has only improved slightly. I have also attached an image to show that the last logical processor is being fully utilized during the simulations just for clarity. I am new to using conda environments and jupyter notebook, but if I find anything that improves performance I'll share it.

![cpu_use](https://github.com/loganbvh/py-tdgl/assets/34236089/73625529-9cdc-41dd-ae4c-03fbc3df965a)

**Comment by loganbvh** [comment]
Most of the time during a simulation is taken up solving a large sparse system using [scipy.sparse.linalg.SuperLU](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.SuperLU.html), which happens once per time step. (You can verify this by using a profiler like https://jiffyclub.github.io/snakeviz/.) `scipy.sparse.linalg.SuperLU` uses the [SuperLU library](https://github.com/xiaoyeli/superlu), which is single-threaded. There is a multi-threaded version of SuperLU (https://github.com/xiaoyeli/superlu_mt), but it has not been integrated into scipy (see this issue https://github.com/scipy/scipy/issues/14096).

I am not an expert in CPU benchmarks, but based on this https://nanoreview.net/en/cpu-compare/intel-core-i9-9900k-vs-apple-m1-pro, the Apple M1 Pro has about 25% better single-core performance than the i9-9900K, so perhaps the difference is not too much of a surprise.

In general, the simulation speed can vary quite a bit depending on the details of the mesh boundary. I think that mesh elements on the boundary can strongly affect the [condition number](https://en.wikipedia.org/wiki/Condition_number) of the linear system that needs to be solved at each time step, and hence the solver speed. For your own models, you can try to slightly increase the number of points in the polygons defining the film's shape (e.g., `polygon.points = polygon.resample(int(1.5 * len(film.points)))` to increase the number of points by 50%). In some cases, this can make a big difference in performance.

**Comment by Kyroba** [comment]
Would changing to a multi-threaded SuperLU be difficult? Would it be as simple as installing the package and then overwriting every call of SuperLU with the multi-threaded version?

Output of `tdgl.version_table()`

![image](https://github.com/loganbvh/py-tdgl/assets/34236089/6313dfe1-413e-414c-9cf3-107ab1ecc318)

I wasn't sure if I should start another issue or comment it here. I am also trying to simulate a 1x1 um YBCO square in applied field no current with parameters:
```
length_units = "um"
# Material parameters
xi = 0.002
london_lambda = 0.1
d = 0.2
layer = tdgl.Layer(coherence_length=xi, london_lambda=london_lambda, thickness=d, gamma=1)

# Device geometry
total_width = 1
total_length = 1

film = (
    tdgl.Polygon("film", points=box(total_width, total_length))
    .resample(1001)
    .buffer(0)
)

film.points = film.resample(int(1.5 * len(film.points)))

```

Using a coherence length this small for a large geometry results in extremely long simulation times I am assuming due to the density of the mesh. Have you done many large scale simulations using parameters like this? I would imagine that this needs to be changed:

`device.make_mesh(max_edge_length=xi / 2, smooth=100)
`

where the `max_edge_length=xi / 2` could be much larger to reduce simulation time but still be accurate.

**Comment by loganbvh** [comment]
> Would changing to a multi-threaded SuperLU be difficult? Would it be as simple as installing the package and then overwriting every call of SuperLU with the multi-threaded version?

It is a bit complicated because the SuperLU library has to be compiled during the scipy build process. As of a couple of years ago, someone was working on it (https://github.com/scipy/scipy/issues/14096), so you could comment on that issue to indicate your support for the feature.

> Using a coherence length this small for a large geometry results in extremely long simulation times I am assuming due to the density of the mesh. Have you done many large scale simulations using parameters like this?

Yes, it is possible to simulate models where the mesh has hundreds of thousands of sites, but of course it begins to take quite a long time. If the applied field is small enough that you do not expect any vortices to be generated, you could use a longer `max_edge_length`. I think that results will generally not be accurate if you have vortices and use `max_edge_length` significantly greater than `xi / 2`, but you could try it.

In any case, the device you are trying to simulate violates two important assumptions of `tdgl`:

1. The film thickness must be smaller than or comparable to the coherence length so that the order parameter is uniform in the out-of-plane (z) direction.
2. The film size is smaller than or comparable to the effective penetration depth `Lambda = london_lambda**2 / d` (in at least one direction) so that screening is negligible.

As a result, even if the simulations were faster to run, the results would probably not be physically meaningful. If you are interested in applied fields (currents) that are small compared to the critical field (current), you could check out https://github.com/loganbvh/superscreen, which solves the 2D London equation and self-consistently models Meissner screening. The downside is that the London model is a linear model, so it does not capture the suppression of the superfluid density by large applied fields or currents.

**Comment by Kyroba** [comment]
Thanks for the advice! I'm thinking of using a simpler shape for my experiment that still fits within the guidelines you mentioned. Even if the specific properties of the material don't match YBCO exactly, I could adjust other factors like coherence length and penetration depth to similar ratios. This would help me understand how vortices behave in a way that's similar to what I actually want to study, but easier to simulate.

**Comment by loganbvh** [comment]
One other thing to mention is that because the solver uses an adaptive time step, the predicted simulation time displayed in the progress bar near the beginning of a simulation is usually a significant over-estimate. In other words, the first half of the progress bar may take much longer than the second half because the time step is shorter at earlier times.

This is especially true if you are running a model with an applied field but no transport current. As vortices first nucleate, the solver must use a small time step to resolve the vortex motion (so the total time predicted in the progress bar will be very long). However, as the vortices reach their equilibrium position, the solver can use a much larger time step and the simulation goes more quickly. So I would recommend trying a "full-size" simulation and leaving it to run for a while (for example, overnight) to see how long it actually takes.

**Comment by Kyroba** [comment]
I will give it a go tonight and share the results. Also, here is a simulation of just a square geometry in an applied field of 500mT. 

https://github.com/loganbvh/py-tdgl/assets/34236089/eed85e15-0a5c-463c-83be-6c3ed39f9579

My intuition was that the vortices would nucleate from the corner sites, but it is not the case here. Does this seem like a mesh generation issue to you? Perhaps increasing the points via `polygon.resample` would improve it.

**Comment by loganbvh** [comment]
The animation looks very reasonable to me. The vortices will not nucleate in the corners because the Meissner current density vanishes in the corners (you can verify this by also animating the supercurrent density)

**Comment by Kyroba** [comment]
I believe you are right, I have only seen how physical defects and mesoscopic corners (which would most likely introduce a lot of defects) nucleate the vortices in experiments. Looking at various other simulations done online and the supercurrent density, it is clear that they should nucleate from the edges.

Thank you very much!

**Comment by Kyroba** [comment]
I was wondering if it is possible to use inhomogenous applied vector fields such as gradients? Looking at the parameter in the documentation it does seem like you should be able to, but how could I implement this?

applied_vector_potential ([Union](https://docs.python.org/3/library/typing.html#typing.Union)[[Callable](https://docs.python.org/3/library/typing.html#typing.Callable), [float](https://docs.python.org/3/library/functions.html#float)]) – A function or [tdgl.Parameter](https://py-tdgl.readthedocs.io/en/latest/api/solver.html#tdgl.Parameter) that **_computes the applied vector potential as a function of position (x, y, z)_**. If a float B is given, the applied vector potential will be that of a uniform magnetic field with strength B field_units.

For instance, is it possible to have the left side of the square have positive applied field and the other have negative? This may be interesting to observe vortex movement in the proximity of a ferromagnetic Néel wall.

I hope I didn't just brush over something in the documentation but it does not seem obvious how I could simulate something like this. There is the union argument, so I am assuming it works similarly to how the device itself is created via polygons?

**Comment by loganbvh** [comment]
You would have to figure out how to represent the vector potential corresponding to the magnetic field distribution that you want. One relatively easy case is the vector potential from a current distribution: https://en.wikipedia.org/wiki/Magnetic_vector_potential#Calculation_of_potentials_from_source_distributions. For example, if you had a current carrying wire parallel to the film, the magnetic field from the current would be positive on one side of the wire and negative on the other side of the wire. Maybe a series of such wires stacked in parallel along the z axis (or a sheet of current parallel to the z axis) could approximate the magnetic field distribution you are looking for?

Some examples of how to define the applied vector potential are given in https://github.com/loganbvh/py-tdgl/tree/main/tdgl/sources

**Comment by loganbvh** [comment]
@Kyroba, I am thinking of adding support for other sparse solvers in addition to SuperLU. Scipy can also use the UMFPACK solver if [SuiteSparse](https://people.engr.tamu.edu/davis/suitesparse.html) and [scikit-umfpack](https://pypi.org/project/scikit-umfpack/) are installed. On my MacBook, it seems like SuperLU is faster than UMFPACK, but it may depend on the platform and specific problem.

I can also add the option to use the Intel PARDISO solver (via [pypardiso](https://pypi.org/project/pypardiso/)), which is multithreaded and may be faster than both SuperLU and UMFPACK on Intel CPUs. Would you be interested in helping to test py-tdgl with the PARDISO solver on Windows with Intel CPU?

**Comment by Kyroba** [comment]
I'd be more than happy to help. Interested to see the performance enhancement from PARDISO.

**Comment by loganbvh** [comment]
Awesome! I think the steps will be:

1. Clone the py-tdgl GitHub repo and checkout the `sparse-solvers` branch:
```
git clone --branch sparse-solvers https://github.com/loganbvh/py-tdgl.git
```
2. Create and activate a new conda environment for testing so that you don't mess up your existing env
3. Install the GitHub version of `tdgl` in this new environment:
```
conda activate <env-name>
cd py-tdgl # navigate into the py-tdgl repo
# Locally install py-tdgl from the GitHub repo, including the optional dependency on pypardiso
pip install -e ".[pardiso]"
```
4. Run a simulation with both `options.sparse_solver = "superlu"` and `options.sparse_solver = "pardiso"` and see what happens.

If you want to also try the UMFPACK solver, you will have to install `suitesparse` using conda and then install `scikit-umfpack` using pip:
```
conda install -c conda-forge suitesparse
pip install swig scikit-umfpack
```
Then you can try a simulation with `options.sparse_solver = "umfpack"`.

Thanks for your help and let me know if you run into any issues. It will be interesting to see how the different solvers perform.

**Comment by Kyroba** [comment]
Here are the results using the different parsers for the quickstart geometry.

![image](https://github.com/loganbvh/py-tdgl/assets/34236089/0cbc0897-f6f9-4253-93e4-54cc7417026e), ![image](https://github.com/loganbvh/py-tdgl/assets/34236089/f568c215-2c4f-474f-9d6d-8303262b5dcb)

```
options = tdgl.SolverOptions(
    # Allow some time to equilibrate before saving data.
    sparse_solver="pardiso",
    skip_time=100,
    solve_time=150,
    output_file=os.path.join(tempdir.name, "weak-link-zero-field.h5"),
    field_units = "mT",
    current_units="uA",
    save_every=100,
)
```

### Using PARDISO:
Thermalizing: 100%|███████████████████████████████████████████████████████████████▉| 100/100 [00:28<00:00,  3.56tau/s ]
Simulating: 100%|█████████████████████████████████████████████████████████████████▉| 150/150 [01:13<00:00,  2.04tau/s ]

I even changed the threads used:
```
os.environ["OPENBLAS_NUM_THREADS"] = "8"
```
but this did not impact the result at all:

Thermalizing: 100%|███████████████████████████████████████████████████████████████▉| 100/100 [00:29<00:00,  3.39tau/s ]
Simulating: 100%|██████████████████████████████████████████████████████████████████| 150/150 [01:13<00:00,  2.03tau/s ]

### Using SuperLU:
Thermalizing: 100%|███████████████████████████████████████████████████████████████▉| 100/100 [00:18<00:00,  5.42tau/s ]
Simulating: 100%|█████████████████████████████████████████████████████████████████▉| 150/150 [00:45<00:00,  3.27tau/s ]

SuperLU has better performance here.

This may be expected as a single threaded computation could outperform multi-threaded for smaller geometries as there is overhead for PARDISO. I am now simulating a larger geometry and will send through the results soon.

![image](https://github.com/loganbvh/py-tdgl/assets/34236089/0bf54491-c3f5-46e3-942a-1fb48b888ad3)

Windows Task Manager also shows appropriate CPU usage during PARDISO solving.

**Comment by Kyroba** [comment]
These are the outputs for a larger system.
![image](https://github.com/loganbvh/py-tdgl/assets/34236089/2645a667-5809-42f5-90d3-c6f44f7c7399), ![image](https://github.com/loganbvh/py-tdgl/assets/34236089/8fe8a760-bd32-4cbe-a62c-a7d7f1f45e48)

### Using PARDISO:
Simulating: 100%|█████████████████████████████████████████████████████████████▉| 1000/1000 [2:33:54<00:00,  9.24s/tau ]

### Using SuperLU:
Simulating: 100%|█████████████████████████████████████████████████████████████▉| 1000/1000 [2:47:34<00:00, 10.05s/tau ]

The performance boost seems negligible. Maybe the bottleneck is somewhere else? I can also try UMFPACK tomorrow.

**Comment by loganbvh** [comment]
Interesting results. It looks like there are some inefficiencies in pypardiso.spsolve (they hash the entire sparse matrix each time spsolve is called to see whether it has already been factorized). This may slow things down given how many times spsolve must be called by py-tdgl. I can try to change how pypardiso is called from py-tdgl to avoid this unnecessary extra work. I am traveling this weekend, but I will try those changes next week.

**Comment by loganbvh** [comment]
One other thing that you could try is to run `tdgl` using Windows Subsystem Linux (WSL, https://learn.microsoft.com/en-us/windows/wsl/install). Instructions on how to install Anaconda in WSL: https://gist.github.com/kauffmanes/5e74916617f9993bc3479f401dfec7da. It is possible that SuperLU is simply highly optimized for Linux (and other unix-like OS)

**Comment by Kyroba** [comment]
This is for the Quickstart simulating with zero applied field:
![image](https://github.com/loganbvh/py-tdgl/assets/34236089/612596ab-b141-4f54-88b8-f6c5c7d9d3f4)

**Comment by Kyroba** [comment]
> Some examples of how to define the applied vector potential are given in https://github.com/loganbvh/py-tdgl/tree/main/tdgl/sources. A `tdgl.Parameter` is simply a way to package a function with some default arguments; you can also just write a plain Python function that calculates the vector potential.

Looking deeper into the functions `current_loop_vector_potential` and `current_loop_field`, they look very similar. I assume we use  `current_loop_vector_potential`  like so: 
```
from tdgl.sources import CurrentLoop

options = tdgl.SolverOptions(
    solve_time=400,
    output_file=os.path.join(tempdir.name, "weak-link-zero-current-loop.h5"),
    field_units = "mT",
    current_units="uA",
)

loop_parameter = CurrentLoop(
    current=10000,  # Set your current value here
    radius=0.5,  # Set your loop radius here
    center=(0, 0, 0.3),  # Set the center position of your loop here
    length_units="um",   # Set your length unit, default is "um"
)

zero_current_solution = tdgl.solve(
    device,
    options,
    # If applied_vector_potential is given as a single number,
    # it is interpreted to mean the vector potential associated with a
    # uniform out-of-plane magnetic field with the specified strength.
    applied_vector_potential= loop_parameter, # optionally can include a uniform field too via + [B]
)
```
What would be the use of `current_loop_field`?
Also the applied_vector_potential output looks like this for the above system:
![image](https://github.com/loganbvh/py-tdgl/assets/34236089/4090604b-1359-476c-a7aa-befde2488ac6)
The loop is not well defined, is this a segmenting issue of the loop?

**Comment by Kyroba** [comment]
Sorry to keep extending this issue, I am not sure if I should just start a new one. I am trying to implement a custom vector potential but I am having issues understanding what is exactly passed into the solver. So for example, I wanted to try to put an infinite current carrying wire perpendicular to the film. The magnetic fields induced would be in the plane of the film. This won't necessarily interact with the vortices but I want to see if it can apply an asymmetric force on the vortices when applied in conjunction with the loop vector potential.

Ok, so we can use Ampere's law to determine the magnetic fields produced. Visualizing in python:
![image](https://github.com/loganbvh/py-tdgl/assets/34236089/ba323ed2-11b5-4e48-abca-b8b79df03026)

Now, the vector potential A would be a vector field all pointing parallel to the wire with magnitudes proportional to ln(1/r), r as the radial distance from the wire. Is it as simple as inputting this vector field I externally calculated into the solver's `applied_vector_potential = ...`? I think I need to take into account scaling terms, but not sure how I can implement this.

I might just copy everything in the function `current_loop_vector_potential` and create a new one with:

```
    # r_0 is some small value to prevent division by 0 
    mag = mu_0 * current  / (2 * np.pi) * np.log(1/(rs+r_0))
    # \vec{A} is directed along the axial direction,
    # so here we generate the axial unit vector.
    direc = np.array([0,0,1])
    return mag * direc * ureg("T * m")
```

I think rs in the above is just from:
`positions = positions - loop_center`
`rs = np.linalg.norm(positions, axis=1)` but since we only care about the radial distance:
`rs = np.linalg.norm(positions[:,:2], axis=1)` 


Then call this by also adding a wire.py file in sources similar to how constant and loop are done, then importing it in  `__init__.py`

I'm not an experienced developer so I am not really sure how all these files link together, but I hope the logic is sound.

**Comment by loganbvh** [comment]
Because `tdgl` solves a two-dimensional TDGL model, there is no way to simulate a purely in-plane magnetic field. Only the z component of the magnetic field (i.e., the x and y components of the magnetic vector potential) can interact with the order parameter. The z component of the applied vector potential is ignored completely in the solver: https://github.com/loganbvh/py-tdgl/blob/db107085c1da4bff4fe8d7f14dba9c02fc90c490/tdgl/solver/solve.py#L265-L267

I think the output of your function should have shape `(len(rs), 3)`, so the return statement should be

```python
    direc = np.array([[0,0,1]])  # shape (1, 3)
    # change mag from shape (len(rs), ) to shape (len(rs), 1)
    return mag[:, np.newaxis] * direc * ureg("T * m")
```

In terms of scaling, the `applied_vector_potential` function or `Parameter` should return **A** in units of `options.field_units * device.length_units`, so for example you can
```python
return (mag[:, np.newaxis] * direc * ureg("T * m")).to("mT * um").magnitude
````
assuming that `options.field_units = "mT"`, `device.length_units = "um"`, and that `positions` are given in meters and `current` is given in amperes.

> I think rs in the above is just from:

That looks correct to me.

> Then call this by also adding a wire.py file in sources similar to how constant and loop are done, then importing it in __init__.py

That works. You could also just put wire.py in the same directory as your Jupyter notebook or main Python script and import it from there.

**Comment by loganbvh** [comment]
> Does this mean that for instance a magnetic field produced by a current carrying loop with a superimposed in-plane component would have a different output when run in tdgl  compared to without the in-plane component?

No, sorry for the confusing phrasing. The z component of the vector potential is completely ignored, so in-plane fields do not exist in `tdgl`.

**Comment by loganbvh** [comment]
Closing this issue because I have merged https://github.com/loganbvh/py-tdgl/pull/35 which adds the umfpack and pardiso solvers. Feel free to reopen this issue or open a new issue if you have any other questions/suggestions.
